src/networks.jl

```diff
@@ -994,3 +994,262 @@ Carries on the spatial convolution respecting periodicity at the boundaries.
+              
+              # Decoding
+              gnorm4=GroupNorm(channels[4], 32, swish),
+              tconv4=Upsampling(channels[4] => channels[3], nspatial, kernel_size=inner_kernelsize, periodic=periodic),
+              denset4=Dense(embed_dim, channels[3]),
+              control_denset4=Dense(embed_dim, channels[3]),
+              tgnorm4=GroupNorm(channels[3], 32, swish),
+              
+              tconv3=Upsampling(channels[3]+channels[3] => channels[2], nspatial, kernel_size=middle_kernelsize, periodic=periodic),
+              denset3=Dense(embed_dim, channels[2]),
+              control_denset3=Dense(embed_dim, channels[2]),
+              tgnorm3=GroupNorm(channels[2], 32, swish),
+              
+              tconv2=Upsampling(channels[2]+channels[2] => channels[1], nspatial, kernel_size=outer_kernelsize, periodic=periodic),
+              denset2=Dense(embed_dim, channels[1]),
+              control_denset2=Dense(embed_dim, channels[1]),
+              tgnorm2=GroupNorm(channels[1], 32, swish),
+              
+              # Projection
+              tconv1=tconv1,
+              mean_bypass_layers...
+              )
+
+    return ControlledNoiseConditionalScoreNetwork(layers, control_net, mean_bypass, scale_mean_bypass, shift_input, shift_output, gnorm)
+end
+
+@functor ControlledNoiseConditionalScoreNetwork
+
+"""
+    (net::ControlledNoiseConditionalScoreNetwork)(x, c, t)
+
+Evaluates the neural network of the NoiseConditionalScoreNetwork
+model on (x,c,t), where `x` is the tensor of noised input,
+`c` is the tensor of contextual input, and `t` is a tensor of times.
+"""
+function (net::ControlledNoiseConditionalScoreNetwork)(x, c, t)
+    # Get size of spatial dimensions
+    nspatial = ndims(x) - 2
+
+    # Embeddings
+    embed = net.layers.gaussfourierproj(t)
+    embed = net.layers.linear(embed)
+    control_embed = net.control_net(c)
+
+    # Encoder
+    if net.shift_input
+        h1 = x .- mean(x, dims=(1:nspatial)) # remove mean of noised variables before input
+    else
+        h1 = x
+    end
+    h1 = net.layers.conv1(h1)
+    h1 = h1 .+ expand_dims(net.layers.dense1(embed) .+ net.layers.control_dense1(control_embed), nspatial)
+    h1 = net.layers.gnorm1(h1)
+    h2 = net.layers.conv2(h1)
+    h2 = h2 .+ expand_dims(net.layers.dense2(embed) .+ net.layers.control_dense2(control_embed), nspatial)
+    h2 = net.layers.gnorm2(h2)
+    h3 = net.layers.conv3(h2)
+    h3 = h3 .+ expand_dims(net.layers.dense3(embed) .+ net.layers.control_dense3(control_embed), nspatial)
+    h3 = net.layers.gnorm3(h3)
+    h4 = net.layers.conv4(h3)
+    h4 = h4 .+ expand_dims(net.layers.dense4(embed) .+ net.layers.control_dense4(control_embed), nspatial)
+
+    # middle
+    h = h4
+    for block in net.layers.resnet_blocks
+        h = block(h, embed .+ control_embed) # add in control embedding, can perhaps be done better.
```

Comment:
My only thought, which you probably already considered, would be to add another dense layer to the ResnetBlockNCSN here:
https://github.com/CliMA/CliMAgen.jl/blob/7b50239299059f94a2733a7e6c5e5ecf311b2e8c/src/networks.jl#L810

and do the same thing that we do for time, but I think this is a fine place to start. not sure it'd make much of a difference.

Reply:
(and pass in control_embed as a third arg)

Reply:
Let's see if we can get it to run and do something and then add this.